contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,244 @@
+ * Contract for minting and redeeming any Set token using
+ * ETH as the paying/receiving currency. All swaps are done using the best price
+ * found on Uniswap or Sushiswap.
+ *
+ */
+contract ExchangeIssuance {
+
+    using SafeMath for uint256;
+    
+    /* ============ State Variables ============ */
+
+    IUniswapV2Router02 private uniRouter;
+    address private uniFactory;
+    IUniswapV2Router02 private sushiRouter;
+    address private sushiFactory;
+
+    IBasicIssuanceModule private basicIssuanceModule;
+    address private WETH;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(address indexed recipient, address indexed index, address indexed inputToken, uint256 amount);
+    event ExchangeRedeem(address indexed recipient, address indexed index, address indexed outputToken, uint256 amount);
+
+    /* ============ Constructor ============ */
+
+    constructor(address _uniFactory, address _uniRouter, address _sushiFactory, address _sushiRouter, address _basicIssuanceModule) public {
+        uniFactory = _uniFactory;
+        uniRouter = IUniswapV2Router02(_uniRouter);
+
+        sushiFactory = _sushiFactory;
+        sushiRouter = IUniswapV2Router02(_sushiRouter);
+
+        WETH = uniRouter.WETH();
+        basicIssuanceModule = IBasicIssuanceModule(_basicIssuanceModule);
+        IERC20(WETH).approve(address(uniRouter), PreciseUnitMath.maxUint256());
+        IERC20(WETH).approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+    }
+
+    /* ============ External Functions ============ */
+
+    receive() external payable {}
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming an index. This function must only be called before the first time
+     * this smart contract is used on any particular index, or when a new token is added
+     * to the index.
+     *
+     * @param _index    Address of the index being initialized
+     */
+    function initApprovals(address _index) external {
+        ISetToken.Position[] memory positions = ISetToken(_index).getPositions();
+        for (uint256 i=0; i<positions.length; i++) {
+            IERC20 token = IERC20(positions[i].component);
+            token.approve(address(uniRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(basicIssuanceModule), PreciseUnitMath.maxUint256());
+        }
+    }
+
+    /**
+     * Redeems an Index and sells the underlying tokens using Uniswap
+     * or Sushiswap.
+     *
+     * @param _index        Address of the index being redeemed
+     * @param _amount       The amount of the index to redeem
+     * @param _isOutputETH  Set to true if the output token is Ether
+     * @param _outputToken  Address of output token. Ignored if _isOutputETH is true
+     */
+    function exchangeRedeem(address _index, uint256 _amount, bool _isOutputETH, address _outputToken) external {
```

Comment:
To protect users from unexpected slippage, we should enforce a `minReceive` on how much ERC20/ETH the user gets back. This would be similar to how `exchangeIssue` has a max spend and how uniswap allows input of a minReceive quantity

Reply:
+1. A swap without appropriate `minReceive` and `maxSpent` might lead to a sandwich attack by a third party.